rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,302 @@
+      "username": "ainspector",
+      "badgeNumber": "123"
+    },
+    "tripKey": {
+      "service-date": "2022-01-20",
+      "id": "ADDED-1"
+    },
+    "stop": "place-lake",
+    "time": "10:00:00",
+    "tripType": "round",
+    "consist": [
+      {"carNumber": "3850"}
+    ],
+    "operators": [
+      {"badgeNumber": "567"}
+    ]
+  }
+}
+```
+
+In this instance, the inspector dropped the 10:00am trip, and created a new trip with the same departure time. Ideally, RTR will treat this the same as if the 10:00am were updated to be a 1-car trip, but inspectors do not always enter the data in that fashion.
+
+# Drawbacks
+- This requires Glides to include new functionality to write their messages into a Kinesis stream. However, they already use `ex_aws` so additional messages should not be hard to add.
+ 
+# Rationale and alternatives
+- RTR already consumes a Kinesis stream of CloudEvents from OCS, 
+- so it's a limited additional lift to consume a new Kinesis stream
+- Kinesis Firehose allows writing the stream of messages to S3 for LAMP to consume, without needing a separate integration
+
+## Alternative: split update types into different events
+Examples:
+- set departure time
+- set/unset operator
+- set/unset consist
+
+The current trainsheet interface does not distinguish between these cases, so it would need to be added to Glides.
+
+## Alternative: departure time as RFC3999 string
+This would be easy to parse with standard tools, but doesn't match existing OCS behavior which only provides a time.
+
+## Alternative: GTFS-RT ServiceChange feed
+For much of CTD's history, we have used versions of GTFS-RT as our interchange format between applications. Current examples include:
+- Operator assignments and block waivers (Busloc -> Skate)
+- Arrival predictions at skipped stops (RTR -> Realtime Signs)
+
+If this data can be directly presented to riders, this avoids a processing step.
+
+However, this is also limiting:
+- both sides must already understand GTFS
+- can be hard to reproduce user intent
+- expects full (or at least mostly full) records
+
+While the ladder view in Glides understands GTFS-RT, the trainsheet information comes directly from HASTUS on a quarterly basis. It does not have updated trip IDs, so RTR would need to process the data anyways: it could not be provided directly to riders.
+# Unresolved questions
+- While some sample CloudEvent formats are included here, the specific format should be settled on as a conversation between the Glides and Transit Data teams.
+- Some trainsheet entries (removing an operator from their trips) can happen the day before. Do we need to persist any events longer than the 24 hour Kinesis default?
+# Future possibilities
+- As Glides becomes more widely adopted, further information entered by Chief Inspectors and other officials can be included as events, such as short turns, express trips, and non-revenue trips. However, this is not currently scheduled for implementation.
```

Comment:
Even if we don't include them in this RFC, I think it'd be useful to write out a plan for how they'll happen. Adding fields to `trip-updated`? New events? Will the event version numbers need to get bumped?

Reply:
I think any of those are fine implementations, but which one is the most appropriate depends on what the data actually looks like at the time, and what other data (such as ODS) is available.

Reply:
I don't just mean what the data format is. I mostly mean process stuff like:
- Do we need a new RFC to change the format?
- Should consumers safely ignore unknown fields, so that Glides can add new fields to existing events?
- Under what situations would the version number increase?

I want to avoid the situation where we finalize this RFC, and then want to add a feature, and the path to do that is unclear or difficult.

Reply:
Those are good questions; I'll add responses to the reference section.